store/src/java/com/zimbra/cs/service/admin/Auth.java

```diff
@@ -74,378 +56,115 @@
-
-		if (registerTrustedDevice && (trustedToken == null || trustedToken.isExpired())) {
-			// generate a new trusted device token if there is no existing one or if the
-			// current one is no longer valid
-			Map<String, Object> attrs = getTrustedDeviceAttrs(zsc, newDeviceId == null ? deviceId : newDeviceId);
-			if (trustedDeviceManager != null) {
-				trustedToken = trustedDeviceManager.registerTrustedDevice(attrs);
-			}
-		}
-
-		return null;
-	}
-
-    private Map<String, Object> getTrustedDeviceAttrs(ZimbraSoapContext zsc, String deviceId) {
-        Map<String, Object> deviceAttrs = new HashMap<String, Object>();
-        deviceAttrs.put(AuthContext.AC_DEVICE_ID, deviceId);
-        deviceAttrs.put(AuthContext.AC_USER_AGENT, zsc.getUserAgent());
-        return deviceAttrs;
-    }
-    
-    private void verifyTrustedDevice(Account account, TrustedDeviceToken td, Map<String, Object> attrs) throws ServiceException {
-        TrustedDevices trustedDeviceManager = TwoFactorAuth.getFactory().getTrustedDevices(account);
-        trustedDeviceManager.verifyTrustedDevice(td, attrs);
-    }
-
-    private Element needTwoFactorAuth(Account account, TwoFactorAuth auth, ZimbraSoapContext zsc, TokenType tokenType, Map<String, Object> context, boolean csrfSupport) throws ServiceException {
-        /* two cases here:
-         * 1) the user needs to provide a two-factor code.
-         *    in this case, the server returns a two-factor auth token in the response header that the client
-         *    must send back, along with the code, in order to finish the authentication process.
-         * 2) the user needs to set up two-factor auth.
-         *    this can happen if it's required for the account but the user hasn't received a secret yet.
-         */
-        if (!auth.twoFactorAuthEnabled()) {
-            throw AccountServiceException.TWO_FACTOR_SETUP_REQUIRED();
-        } else {
-            Element response = zsc.createElement(AccountConstants.AUTH_RESPONSE);
-            AuthToken twoFactorToken = AuthProvider.getAuthToken(account, Usage.TWO_FACTOR_AUTH, tokenType);
-            response.addUniqueElement(AccountConstants.E_TWO_FACTOR_AUTH_REQUIRED).setText("true");
-            response.addAttribute(AccountConstants.E_LIFETIME, twoFactorToken.getExpires() - System.currentTimeMillis(), Element.Disposition.CONTENT);
-            twoFactorToken.encodeAuthResp(response, false);
-            response.addUniqueElement(AccountConstants.E_TRUSTED_DEVICES_ENABLED).setText(account.isFeatureTrustedDevicesEnabled() ? "true" : "false");
-
-            HttpServletRequest httpReq = (HttpServletRequest) context.get(SoapServlet.SERVLET_REQUEST);
-            HttpServletResponse httpResp = (HttpServletResponse) context.get(SoapServlet.SERVLET_RESPONSE);
-
-            setCSRFToken(httpReq, httpResp, twoFactorToken, csrfSupport, response);
-
-            return response;
-        }
-    }
-
-    private void checkAdmin(Account acct) throws ServiceException {
```

Comment:
Do not remove this method from here as it will break the existing implementation of the clients. You can use override or overload but do **NOT** remove the existing methods from the Zimbra organisation codebase.

Reply:
This is a private method and its scope is inside of this class.  We can not override this method as it is private.
It is used to check that a specific account object is Admin Account or not. So I believed this method can be reused and moved to utility class. If still there is a gap in my understanding then please let me know.

Reply:
How do you know nobody is using the same method within the same class where it's scope is inside of this class.

Reply:
As this is a private method, checked that this is used from two places only in Admin/Auth.java. Prashant had also suggested this kind of change.

Reply:
Zimbra is open source and can be used by anyone how can you make sure someone has not referenced any of these methods in the same class.

Reply:
Added methods in file.